README.md
# fetch-readablestream
Compatibility layer for efficient streaming of binary data using [WHATWG Streams](https://streams.spec.whatwg.org/)

## Why
This library provides a consistent, cross browser API for streaming a response from an HTTP server based on the [WHATWG Streams specification](https://streams.spec.whatwg.org/).  At the time of writing, Chrome is the only browser to nativley support returning a `ReadableStream` from it's `fetch` implementation - all other browsers need to fall back to `XMLHttpRequest`.

FireFox does provide the ability to efficiently retrieve a byte-stream from a server; however only via it's `XMLHttpRequest` implementation (when using `responsetype=moz-chunked-arraybuffer`).  Other browsers do not provide access to the underlying byte-stream and must therefore fall-back to concatenating the response string and then encoding it into it's UTF-8 byte representation using the [`TextEncoder` API](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder).

*Nb:* If you are happy using a node-style API (using callbacks and events) I would suggest taking a look at [`stream-http`](https://github.com/jhiesey/stream-http).

## Installation
This package can be installed with `npm`:

```
$ npm install fetch-readablestream --save
```

Once installed you can import it directly:

```js
import fetchStream from 'fetch-readablestream';
```

Or you can add a script tag pointing to the `dist/fetch-readablestream.js` bundle and use the `fetchStream` global:

```html
<script src="./node_modules/fetch-readablestream/dist/fetch-readablestream.js"></script>
<script>
  window.fetchStream('...')
</script>
```

## Usage
The `fetchStream` api provides a subset of the [`fetch` API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch); in particular, the ability to get a `ReadableStream` back from the `Response` object which can be used to efficiently stream a chunked-transfer encoded response from the server.

```js
function readAllChunks(readableStream) {
  const reader = readableStream.getReader();
  const chunks = [];

  function pump() {
    return reader.read().then(({ value, done }) => {
      if (done) {
        return chunks;
      }
      chunks.push(value);
      return pump();
    });
  }

  return pump();
}

fetchStream('/endpoint')
  .then(response => readAllChunks(response.body))
  .then(chunks => console.dir(chunks))
```

## Browser Compatibility
`fetch-readablestream` makes the following assumptions on the environment; legacy browsers will need to provide Polyfills for this functionality:

| Feature                        | Browsers                         | Polyfill |
|--------------------------------|----------------------------------|----------|
| ReadableStream                 | Firefox, Safari, IE11, PhantomJS | [web-streams-polyfill](https://www.npmjs.com/package/web-streams-polyfill) |
| TextEncoder                    | Safari, IE11, PhantomJS          | [text-encoding](https://www.npmjs.com/package/text-encoding) |
| Promise, Symbol, Object.assign | IE11, PhantomJS                  | [babel-polyfill](https://www.npmjs.com/package/babel-polyfill) |

## Contributing
Use `npm run watch` to fire up karma with live-reloading.  Visit http://localhost:9876/ in a bunch of browsers to capture them - the test suite will run automatically and report any failures.




CHANGELOG.md
## [0.1.0] - 17/09/2016
- Initial release, lifted from github.com/jonnyreeves/chunked-request


test/integ/util.js
export function drainResponse(response) {
  const chunks = [];
  const reader = response.body.getReader();

  function pump() {
    return reader.read()
        .then(({ done, value }) => {
          if (done) {
            return chunks
          }
          chunks.push(value);
          return pump();
        })
  }
  return pump();
}

export function decodeUnaryJSON(chunks) {
  return new Promise((resolve) => resolve(JSON.parse(decodeResponse(chunks))))
}

function decodeResponse(chunks) {
  const decoder = new TextDecoder();
  return chunks
      .map((bytes, idx) => {
        const isLastChunk = idx === (chunks.length - 1);
        return decoder.decode(bytes, { stream: !isLastChunk })
      })
      .join("");
}

test/server/index.js
'use strict';

const http = require('http');
const url = require('url');
const cookieParser = require('cookie');

// Which port should HTTP traffic be served over?
const httpPort = process.env.HTTP_PORT || 2001;

// How frequently should chunks be written to the response?  Note that we have no
// control over when chunks are actually emitted to the client so it's best to keep
// this value high and pray to the gods of TCP.
//
// Nb: lower values appear to cause problems for internet explorer.
const CHUNK_INTERVAL_MS = process.env.CHUNK_INTERVAL_MS || 1000;

let _lastRequestClosedByClient = false;

function readBody(req) {
  return new Promise(function (resolve) {
    const body = [];
    req.on('data', function (chunk) {
      body.push(chunk);
    }).on('end', function () {
      resolve(Buffer.concat(body).toString('utf8'));
    });
  });
}

function echoHandler(req, res) {
  readBody(req)
      .then(function (body) {
        res.setHeader('x-powered-by', 'nodejs');
        res.end(JSON.stringify({
          headers: req.headers,
          method: req.method,
          cookies: cookieParser.parse(req.headers.cookie || ''),
          body: body
        }));
      });
}

function serveChunksHandler(req, res) {
  _lastRequestClosedByClient = false;
  req.on('close', function () {
    _lastRequestClosedByClient = true;
  });

  readBody(req)
      .then(function (body) {
        try {
          const chunks = JSON.parse(body);

          res.setHeader('Transfer-Encoding', 'chunked');
          res.setHeader('Content-Type', 'text/html; charset=UTF-8');

          for (let i = 0; i < chunks.length; i++) {
            setTimeout(function (idx) {
              res.write(chunks[idx] + "\n");
              if (idx === chunks.length - 1) {
                res.end();
              }
            }, i * CHUNK_INTERVAL_MS, i);
          }
        } catch (e) {
          res.writeHead(400);
          res.end("Invalid JSON payload: " + e.message);
        }
      });
}

function lastRequestClosedHandler(req, res) {
  res.end(JSON.stringify({ value: _lastRequestClosedByClient }));
}

function failureHandler(req, res) {
  res.writeHead(500);
  res.end("Intentional server error");
}

function handleSrv(req, res) {
  const url = req.parsedUrl;
  switch (url.query.method) {
  case "echo":
    return echoHandler(req, res);
  case "500":
    return failureHandler(req, res);
  case "send-chunks":
    return serveChunksHandler(req, res);
  case "last-request-closed":
    return lastRequestClosedHandler(req, res);
  default:
    res.writeHead(400);
    res.end("Unsupported method: ?method=" + url.query.method);
  }
}

function handler(req, res) {
  req.parsedUrl = url.parse(req.url, true);
  if (req.parsedUrl.pathname === '/srv') {
    return handleSrv(req, res);
  }
  res.writeHead(404);
  res.end("Not found, try /srv");
}

console.log("Serving on http://localhost:" + httpPort);
http.createServer(handler).listen(httpPort);

test/integ/chunked-request.spec.js
import fetchStream from '../../src/index';
import { Headers as HeadersPolyfill } from '../../src/polyfill/Headers';
import { drainResponse, decodeUnaryJSON } from './util';

if (!window.Headers) {
  window.Headers = HeadersPolyfill;
}

// These integration tests run through Karma; check `karma.conf.js` for
// configuration.  Note that the dev-server which provides the `/srv`
// endpoint is proxied through karma to work around CORS constraints.

describe('fetch-readablestream', () => {

  it('returns each chunk sent by the server as a Uint8Array of bytes', (done) => {
    return fetchStream('/srv?method=send-chunks', {
      method: 'POST',
      body: JSON.stringify([ 'chunk1', 'chunk2' ])
    })
        .then(drainResponse)
        .then(chunks => {
          expect(chunks.length).toBe(2, '2 chunks received');
          expect(chunks.every(bytes => bytes.BYTES_PER_ELEMENT === 1)).toBe(true, 'Uint8Array');
        })
        .then(done, done);
  });

  it('will cancel the response stream and close the connection', (done) => {
    return fetchStream('/srv?method=send-chunks', {
      method: 'POST',
      body: JSON.stringify([ 'chunk1', 'chunk2', 'chunk3', 'chunk4' ])
    })
        .then(response => {
          const reader = response.body.getReader();
          return reader.read()
              .then(() => reader.cancel())
        })
        .then(() => fetchStream('/srv?method=last-request-closed'))
        .then(drainResponse)
        .then(decodeUnaryJSON)
        .then(result => {
          expect(result.value).toBe(true, 'response was closed by client');
        })
        .then(done, done);
  });

  it('returns a subset of the fetch API', (done) => {
    return fetchStream('/srv?method=echo', {
      method: 'POST',
      body: "hello world",
    })
        .then(result => {
          expect(result.ok).toBe(true);
          expect(result.headers.get('x-powered-by')).toBe('nodejs');
          expect(result.status).toBe(200);
          expect(result.statusText).toBe('OK');
          expect(result.url).toBe('http://localhost:9876/srv?method=echo');
        })
        .then(done, done);
  });

  it('resolves the fetchStream Promise on server error', (done) => {
    return fetchStream('/srv?method=500')
        .then(result => {
          expect(result.status).toBe(500);
          expect(result.ok).toBe(false);
        })
        .then(done, done);
  });

  it('sends the expected values to the server', (done) => {
    return fetchStream('/srv?method=echo', {
      method: 'POST',
      body: "hello world",
      headers: new Headers({ 'x-foo': 'expected' })
    })
        .then(drainResponse)
        .then(decodeUnaryJSON)
        .then(res => {
          expect(res.method).toBe("POST");
          expect(res.body).toBe("hello world");
          expect(res.headers).toEqual(jasmine.objectContaining({ 'x-foo': 'expected' }));
        })
        .then(done, done);
  });

  it('includes cookies when credentials=include', (done) => {
    document.cookie = 'myCookie=myValue';

    return fetchStream('/srv?method=echo', {
      method: 'POST',
      credentials: 'include'
    })
        .then(drainResponse)
        .then(decodeUnaryJSON)
        .then(res => {
          expect(res.cookies).toEqual(jasmine.objectContaining({ myCookie: "myValue" }));
        })
        .then(done, done);
  });

  /* xhr will always send cookies to the same domain.
  it('omits cookies when credentials=omit', (done) => {
    document.cookie = 'myCookie=myValue';

    return fetchStream('/srv?method=echo', {
      method: 'POST',
      credentials: 'omit'
    })
        .then(drainResponse)
        .then(decodeUnaryJSON)
        .then(res => {
          expect(res.cookies).not.toEqual(jasmine.objectContaining({ myCookie: "myValue" }));
        })
        .then(done, done);
  });
  */

});


src/xhr.js
import { Headers as HeadersPolyfill } from './polyfill/Headers';

export function makeXhrTransport({ responseType, responseParserFactory }) {
  return function xhrTransport(url, options) {
    const xhr = new XMLHttpRequest();
    const responseParser = responseParserFactory();

    let responseStreamController;
    let cancelled = false;

    const responseStream = new ReadableStream({
      start(c) {
        responseStreamController = c
      },
      cancel() {
        cancelled = true;
        xhr.abort()
      }
    });

    const { method = 'GET' } = options;

    xhr.open(method, url);
    xhr.responseType = responseType;
    xhr.withCredentials = (options.credentials !== 'omit');
    if (options.headers) {
      for (const pair of options.headers.entries()) {
        xhr.setRequestHeader(pair[0], pair[1]);
      }
    }

    return new Promise((resolve, reject) => {
      if (options.body && (method === 'GET' || method === 'HEAD')) {
        reject(new TypeError("Failed to execute 'fetchStream' on 'Window': Request with GET/HEAD method cannot have body"))
      }

      xhr.onreadystatechange = function () {
        if (xhr.readyState === xhr.HEADERS_RECEIVED) {
          return resolve({
            body: responseStream,
            headers: parseResposneHeaders(xhr.getAllResponseHeaders()),
            ok: xhr.status >= 200 && xhr.status < 300,
            status: xhr.status,
            statusText: xhr.statusText,
            url: makeResponseUrl(xhr.responseURL, url)
          });
        }
      };

      xhr.onerror = function () {
        return reject(new TypeError('Network request failed'));
      };

      xhr.ontimeout = function() {
        reject(new TypeError('Network request failed'))
      };

      xhr.onprogress = function () {
        if (!cancelled) {
          const bytes = responseParser(xhr.response);
          responseStreamController.enqueue(bytes);
        }
      };

      xhr.onload = function () {
        responseStreamController.close();
      };

      xhr.send(options.body);
    });
  }
}

function makeHeaders() {
  // Prefer the native method if provided by the browser.
  if (typeof Headers !== 'undefined') {
    return new Headers();
  }
  return new HeadersPolyfill();
}

function makeResponseUrl(responseUrl, requestUrl) {
  if (!responseUrl) {
    // best guess; note this will not correctly handle redirects.
    if (requestUrl.substring(0, 4) !== "http") {
      return location.origin + requestUrl;
    }
    return requestUrl;
  }
  return responseUrl;
}

export function parseResposneHeaders(str) {
  const hdrs = makeHeaders();
  if (str) {
    const pairs = str.split('\u000d\u000a');
    for (let i = 0; i < pairs.length; i++) {
      const p = pairs[i];
      const index = p.indexOf('\u003a\u0020');
      if (index > 0) {
        const key = p.substring(0, index);
        const value = p.substring(index + 2);
        hdrs.append(key, value);
      }
    }
  }
  return hdrs;
}

src/polyfill/Headers.js
// Headers is a partial polyfill for the HTML5 Headers class.
export class Headers {
  constructor(h = {}) {
    this.h = {};
    if (h instanceof Headers) {
      h.forEach((value, key) => this.append(key, value));
    }
    Object.getOwnPropertyNames(h)
        .forEach(key => this.append(key, h[key]));
  }
  append(key, value) {
    key = key.toLowerCase();
    if (!Array.isArray(this.h[key])) {
      this.h[key] = [];
    }
    this.h[key].push(value);
  }
  set(key, value) {
    this.h[key.toLowerCase()] = [ value ];
  }
  has(key) {
    return Array.isArray(this.h[key.toLowerCase()]);
  }
  get(key) {
    key = key.toLowerCase();
    if (Array.isArray(this.h[key])) {
      return this.h[key][0];
    }
  }
  getAll(key) {
    return this.h[key.toLowerCase()].concat();
  }
  forEach(callback, thisArg) {
    Object.getOwnPropertyNames(this.h)
        .forEach(key => {
          this.h[key].forEach(value => callback.call(thisArg, value, key, this));
        }, this);
  }
}

src/defaultTransportFactory.js
import fetchRequest from './fetch';
import { makeXhrTransport } from './xhr';

let selected = null;

export default function defaultTransportFactory() {
  if (!selected) {
    if (window.Response && window.Response.prototype.hasOwnProperty("body")) {
      selected = fetchRequest;
    } else {
      const tmpXhr = new XMLHttpRequest();
      const mozChunked = 'moz-chunked-arraybuffer';
      tmpXhr.responseType = mozChunked;
      if (tmpXhr.responseType === mozChunked) {
        selected = makeXhrTransport({
          responseType: mozChunked,
          responseParserFactory: function () {
            return response => new Uint8Array(response);
          }
        });
      } else {
        selected = makeXhrTransport({
          responseType: 'text',
          responseParserFactory: function () {
            const encoder = new TextEncoder();
            let offset = 0;
            return function (response) {
              const chunk = response.substr(offset);
              offset = response.length;
              return encoder.encode(chunk, { stream: true });
            }
          }
        });
      }
    }
  }
  return selected;
}



karma.conf.js
// Karma configuration
// Generated on Wed Feb 17 2016 15:48:21 GMT+0000 (GMT)

/*eslint-env node*/
/*eslint no-var: 0*/
module.exports = function(config) {

  // Browsers to run on Sauce Labs
  // Check out https://saucelabs.com/platforms for all browser/OS combos
  var customLaunchers = {
    /* Safari appears broken on saucelabs, but works locally.
    'SL_Safari': {
      base: 'SauceLabs',
      browserName: 'safari',
      platform: 'OS X 10.11'
    },
    */
    'SL_Chrome': {
      base: 'SauceLabs',
      browserName: 'chrome',
      platform: 'linux'
    },
    'SL_Firefox': {
      base: 'SauceLabs',
      browserName: 'firefox',
      platform: 'linux'
    },
    /* need to figure out what's broken in edge.
    'SL_Edge': {
      base: 'SauceLabs',
      browserName: 'MicrosoftEdge',
      platform: 'Windows 10'
    },
    */
    'SL_IE10': {
      base: 'SauceLabs',
      browserName: 'internet explorer',
      platform: 'Windows 7',
      version: '11'
    }
  };

  var reporters = ['dots'];
  var browsers = [];
  var singlerun = false;

  if (process.env.SAUCE_USERNAME) {
    reporters.push('saucelabs');
    Array.prototype.push.apply(browsers, Object.keys(customLaunchers));
    singlerun = true;
  }

  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],

    sauceLabs: {
      recordScreenshots: false,
      connectOptions: {
        port: 5757,
        logfile: 'sauce_connect.log'
      },
      public: 'public'
    },

    // list of files / patterns to load in the browser
    files: [
      './node_modules/babel-polyfill/dist/polyfill.js',
      'node_modules/text-encoding/lib/encoding.js',
      'node_modules/web-streams-polyfill/dist/polyfill.js',
      'build/integration-tests.js'
    ],


    // list of files to exclude
    exclude: [
    ],

    proxies: {
      '/srv': 'http://localhost:2001/srv',
    },


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: reporters,


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    browsers: browsers,
    captureTimeout: 120000,
    customLaunchers: customLaunchers,

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: singlerun,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: 1
  })
};
